File: csharp/Day02.cs
namespace advent_of_code_2021;

internal class Day02 : IDay
{
    public int DayNumber => 2;
    public int PartOneTestAnswer => 150;
    public int PartTwoTestAnswer => 900;

    public int ExecutePartOne(string[] input)
    {
        var instructions = input.Select(InputLineToInstruction).ToArray();
        var horizontalPosition = instructions.Where(x => x.Direction == Direction.Forward).Sum(x => x.Steps);
        var downSteps = instructions.Where(x => x.Direction == Direction.Down).Sum(x => x.Steps);
        var upSteps = instructions.Where(x => x.Direction == Direction.Up).Sum(x => x.Steps);
        return horizontalPosition * (downSteps - upSteps);
    }

    public int ExecutePartTwo(string[] input)
    {
        var instructions = input.Select(InputLineToInstruction).ToArray();
        var horizontalPosition = instructions.Where(x => x.Direction == Direction.Forward).Sum(x => x.Steps);
        int depth = 0;
        int aim = 0;
        foreach (var instruction in instructions)
        {
            if (instruction.Direction == Direction.Forward)
            {
                depth += instruction.Steps * aim;
            }
            else
            {
                aim += instruction.Steps * (instruction.Direction == Direction.Down ? 1 : -1);
            }
        }
        return horizontalPosition * depth;
    }

    private static Instruction InputLineToInstruction(string inputLine)
    {
        var parts = inputLine.Split(' ');
        return new Instruction((Direction)Enum.Parse(typeof(Direction), parts[0], true), int.Parse(parts[1]));
    }

    enum Direction
    {
        Forward,
        Down,
        Up,
    }

    readonly record struct Instruction(Direction Direction, int Steps);
}

File: typescript/2021/day02.ts
import { Day } from '../program';

export class Day02 implements Day {
  partOneTestAnswer = 150;
  partTwoTestAnswer = 900;

  executePartOne(input: string[]): number {
    const operations = input.map((x) => {
      const splitString = x.split(' ');
      const direction: string = splitString[0];
      const amount: number = +splitString[1];
      return { direction, amount } as Operation;
    });
    const horizontalPosition = Day02.getSumOfValues(operations, 'forward');
    const down = Day02.getSumOfValues(operations, 'down');
    const up = Day02.getSumOfValues(operations, 'up');
    const depth = down - up;
    return horizontalPosition * depth;
  }

  executePartTwo(input: string[]): number {
    const operations = input.map((x) => {
      const splitString = x.split(' ');
      const direction: string = splitString[0];
      const amount: number = +splitString[1];
      return { direction, amount } as Operation;
    });
    let horizontalPosition = 0;
    let depth = 0;
    let aim = 0;
    for (const operation of operations) {
      switch (operation.direction) {
        case 'forward':
          horizontalPosition += operation.amount;
          depth += aim * operation.amount;
          break;
        case 'down':
          aim += operation.amount;
          break;
        case 'up':
          aim -= operation.amount;
          break;
      }
    }
    return horizontalPosition * depth;
  }

  private static getSumOfValues(
    operations: Operation[],
    direction: string,
  ): number {
    return operations
      .filter((x) => x.direction === direction)
      .map((x) => x.amount)
      .reduce((a, b) => a + b, 0);
  }
}

interface Operation {
  direction: string;
  amount: number;
}
